### PR TITLE
Remove deprecated `optimize` from `Optimizer` (and small other fixes)

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

Remove the deprecated methods from the optimizers and fix the getter of the number of iterations in the optimizer result.

### Details and comments

* `optimize` method is superseded by `minimize`
* SPSA didn't trigger the deprecation warning, so it's included now for the next deprecation cycle
* The `factr` attribute in `P_BFGS` was deprecated 13 months ago
* `OptimizerResult.nit` previously returned the number of Jacobian evaluations instead of the number of iterations
* SPSA didn't correctly set the number of iterations if the optimization was terminated early with the termination checker

### CODE REVIEW

File: releasenotes/notes/remove-deprecated-optimizer-methods-d580a07112ccaa2d.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    deprecation warning was not triggered so the method there is still kept.
+deprecations:
+  - |
+    The :meth:`qiskit.algorithms.optimizers.SPSA.optimize` method is deprecated in 
+    favor of :meth:`qiskit.algorithms.optimizers.SPSA.minimize`, which can be used 
+    as direct replacement. Note that the `minimize` method returns a complete result 
+    object with more information than before available.
+fixes:
+  - |
+    Fix the getter of :meth:`qiskit.algorithms.optimizers.OptimizerResult.nit`, which
```
[the remark this answers is not in the document]
Comment (Cryoris) [reply releasenotes/notes/remove-deprecated-optimizer-methods-d580a07112ccaa2d.yaml]: Done in [8fedffc](https://github.com/Qiskit/qiskit-terra/pull/8123/commits/8fedffc24e1fa66b22a397628210f7527dbed5de) 👍🏻

File: qiskit/algorithms/optimizers/spsa.py [hunk 2]
```diff
@@ -641,14 +646,37 @@ def get_support_level(self):
...
         }
 
+    @deprecate_function(
+        """
+The SPSA.optimize method is deprecated as of Qiskit Terra 0.21.0 and will be removed no
+sooner than 3 months after the release date. Instead, use SPSA.minimize as a replacement,
+which supports the same arguments but follows the interface of scipy.optimize and returns a
+complete result object containing additional information.
+"""
+    )
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply qiskit/algorithms/optimizers/spsa.py:656]: This multi-line string is actually going to have weird spacing and hard line breaks, even if pylint doesn't complain - it's possibly better to make sure it's one line of text (such as the string concatenation example above).  Not a big deal, though.

File: releasenotes/notes/remove-deprecated-optimizer-methods-d580a07112ccaa2d.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  - |
+    Remove the `optimize` method from the :class:`~qiskit.algorithms.optimizers.Optimizer`
+    classes, which is superseded by the :meth:`~.algorithms.optimizers.Optimizer.minimize` method as direct replacement.
+    The one exception is :class:`~qiskit.algorithms.optimizers.SPSA`, where the
+    deprecation warning was not triggered so the method there is still kept.
+deprecations:
+  - |
+    The :meth:`.SPSA.optimize` method is deprecated in
+    favor of :meth:`.SPSA.minimize`, which can be used
+    as direct replacement. Note that the `minimize` method returns a complete result
```
Comment (jakelishman) [inline releasenotes/notes/remove-deprecated-optimizer-methods-d580a07112ccaa2d.yaml:12]:
```suggestion
    as direct replacement. Note that this method returns a complete result
```
(or crossref it, or use two backticks ` ``minimize`` `)

Comment (Cryoris) [reply releasenotes/notes/remove-deprecated-optimizer-methods-d580a07112ccaa2d.yaml]: done in [cf09910](https://github.com/Qiskit/qiskit-terra/pull/8123/commits/cf099109ac17d436446ec7efaf9787ebc94d7d2f)

**Review by mtreinish — DISMISSED** [review]

**Pull request merged**